### PR TITLE
Bigger chunks, faster queries.

### LINKED ISSUE #300: Grow chunks in memory
Chunks are 1k, thickly provisioned. With 10m timeseries this means 10gb ram. 

Chunks will on average be less that half full. A scheme which grows in memory chunks should be able to reclaim 25% of memory pretty easily, maybe more. 

Will explain maths when not on phone.

### LINKED ISSUE #766: Use Prom2 chunk code
Much of the code was converted in #555, but currently Cortex uses chunk code from Prometheus v1.
In v2, chunks are broadly encoded in the same way, with a less-convoluted data structure.  However the Go APIs are rather different.

We'd need to still be able to read v1 chunks for a year minimum, so some wrapping/indirecting layer is required.

### LINKED ISSUE #1045: Bigger chunks
I'd like to build a new chunk format that results in reliably bigger chunks.  Effectively I want to build a "meta chunk" which itself is a series of chunks.  These chunks won't be fixed to 1kb in size, but instead to 12hrs in length.

Any thoughts on this?  Going to experiment today.

If this works I suspect it will solve:
- #300 Grow chunks in memory 
- #448 We're flushing virtually empty chunks in prod
- #766 Use Prom2 chunk code (I'm going to investigate using this for the new code)

### PR DESCRIPTION
**Pull request by tomwilkie**
A bigchunk is a slice of of prometheus/tsdb chunks.  Each individual chunk is fixed at 120 samples; after that we add a new one.  There is no upperbound on the number of samples in a bigchunk.

As part of this PR I've removed a bunch of unused code in the chunk package.

Fixes #1045, fixes #766, fixes #300

### CODE REVIEW

**Comment by bboreham** [comment]
If `pkg/prom1/storage/local/chunk/bigchunk.go` is dealing with prom2 chunks, can it go in a directory reflecting that?
`prom1` should be, as far as possible, the code from Prometheus v1 copied verbatim.

**Comment by tomwilkie** [comment]
> prom1 should be, as far as possible, the code from Prometheus v1 copied verbatim.

I guess you want this so back porting fixes from Prometheus v1 is easier?  Its not really something I'm worried about TBH - there is no development work happening on Prometheus v1, we're more likely to find an fix issues in this package ourself.

> If pkg/prom1/storage/local/chunk/bigchunk.go is dealing with prom2 chunks, can it go in a directory reflecting that?

I started with in in another package, but as I've modelling this as another chunk encoding, there needed to be a bunch of references to it in the chunk package to make it work with the rest of our code.  Therefore I just moved it into this package.

Given that, I was thinking of renaming the directory to `chunkenc` to remove the name clashes with out `chunk` package.  WDYT?

**Comment by bboreham** [comment]
If every chunk in a long-running series is ~12 hours, then every second chunk is indexed in two day-buckets, so we have 50% extra index entries.  I'm beginning to think that day buckets are too short.

**Comment by tomwilkie** [comment]
Looking at the series store - if we have 4 hour chunks, we write 7x3=21
series entries per day. With this we have 9 entries per day. So we write
half the number of entries - 9 vs 21. The number of label entries stays the
same - let’s say 10.

For the chunk store - 4 hour chunks would have 21x10 entries. Big chunks
would have 9x10.

So sure it’s 50% when compared with itself, but when compared to short
chunks i think the amount of extra overlaps you write remains the same -
basically one per day. There is after only one start and end to each day.

**Comment by tomwilkie** [comment]
> My comment was not intended as a criticism of bigger chunks, but a broader comment on the DB design. A 50% overhead is worth talking about no mater what your starting-point is.

Lets take this into a separate issue; agree its a valid concern, but its unrelated to this PR.

**Comment by tomwilkie** [comment]
I've moved the chunk code into `pkg/chunk/encoding` now.

File: pkg/querier/batch/chunk_test.go [hunk 1]
```diff
@@ -31,7 +31,7 @@ func mkChunk(t require.TestingT, from model.Time, points int) chunk.Chunk {
 	metric := model.Metric{
 		model.MetricNameLabel: "foo",
 	}
-	pc, err := promchunk.NewForEncoding(promchunk.DoubleDelta)
+	pc, err := promchunk.NewForEncoding(promchunk.Bigchunk)
```
Comment (gouthamve) [inline pkg/querier/batch/chunk_test.go]: Shouldn't we test for each chunk encoding?

File: pkg/chunk/chunk.go [hunk 2]
```diff
@@ -314,6 +319,10 @@ func (c *Chunk) Decode(decodeContext *DecodeContext, input []byte) error {
 
 	c.encoded = input
 	remainingData := input[len(input)-r.Len():]
+	if int(dataLen) > len(remainingData) {
```
[the remark this answers is not in the document]
Comment (tomwilkie) [reply pkg/chunk/chunk.go]: Yeah I think thats better.

File: pkg/querier/batch/chunk_test.go (hunk 1 above)
Comment (tomwilkie) [reply pkg/querier/batch/chunk_test.go]: We do in pkg/querier/querier_test.go; but will do it here too.
